### ISSUE TITLE
Allow interface as its own construct (not just class modifier)

### ISSUE THREAD

**Issue by mmcdon20** [dart-lang/language#3736]
In dart there is support already for creating interfaces using a class modifier, but the existing support is weird and unintuitive. 

Consider this example:

```dart
abstract interface class Vehicle {
  final String make;
  final String model;
  Vehicle(this.make, this.model);
  
  void moveForward() {
    print('$make $model is moving forward!');
  }
}

class HondaCivic extends Vehicle {
  HondaCivic(): super('Honda', 'Civic');
}
```

For starters, it is a bit of a keyword soup to remember `abstract interface class` whenever you want an `interface`, and you almost never would want just `interface class`. 

It is also weird that
1. you can define class fields in an interface
2. you can define constructors in an interface
3. you can define method bodies in an interface
4. you can extend an interface from within the same library

I propose allowing a separate interface construct that would not allow any of the above.

```dart
interface Vehicle {
  final String make; // error cannot have properties use getter instead;
  final String model; // error cannot have properties use getter instead;
  Vehicle(this.make, this.model); // error cannot have constructor
  
  void moveForward() { // error cannot have method body
    print('$make $model is moving forward!');
  }
}

class HondaCivic extends Vehicle { // error cannot extend interface
  HondaCivic(): super('Honda', 'Civic');
}
```

The errors should guide the user towards using `interface` as intended:

```dart
interface Vehicle {
  String get make;
  String get model;
  void moveForward();
}

class HondaCivic implements Vehicle {
  @override
  final String make = 'Honda';
  @override
  final String model = 'Civic';
  @override
  void moveForward() {
    print('$make $model is moving forward!');
  }
}
```

**Comment by LewisHolliday** [comment]
> For starters, it is a bit of a keyword soup to remember abstract interface class whenever you want an interface, and you almost never would want just interface class.
> 
I'm not sure that constructable interfaces, that is, interfaces that even importing libraries can construct, carry their weight. I think they should be limited such that only the local library can construct them. That's because I think you can sidestep the 'cannot extend' restriction by constructing an instance, then wrapping it in a class that implements its interface, which effectively allows you to inherit its implementation:

```dart
// foo.dart
interface class Foo {
	m(){}
}
```

```dart
// bar.dart
import 'foo.dart';
class Bar implements Foo {
	final foo = Foo();
	m() => foo.m(); // Sort of like extending; just more work!
}
```

However, that isn't quite the same as extending, because you don't inherit, for free, the private implementation/members present in the superclass. But if there's a private interface in the superclass, then why give the superclass the `interface` modifier at all? The author of the superclass is creating potential exceptions for no benefit.

**TLDR**: I think non-abstract interfaces should have no public constructors. I might be way off on that, but I can't think of a good reason for wanting an interface that an outside library can construct. Are there any? If not, that raises another question: should there be a warning or lint against non-abstract `interface class`es with public constructors?

---

<blockquote>

1. you can define class fields in an interface
  
2. you can define constructors in an interface
   
3. you can define method bodies in an interface
   
4. you can extend an interface from within the same library

I propose allowing a separate interface construct that would not allow any of the above.
</blockquote>

I can kind of see where that is coming from. It is definitely weird; especially point 2. But I'm not sure that we should have more than one way to do the same thing. I think these are my issues:
1. `abstract interface class` is too wordy
2. It's too easy to write `interface class`, which allows _outside_ libraries to construct it
3. Nobody should want to construct an `interface class` except the inside library

However, I think there _is_ value in allowing an interface to be constructed and extended from within the same library. And, more importantly, I think there is value in all interfaces being potentially extendable from within the same library. Right now, an author can add an implementation to their `abstract? interface class`, which allows extending. Is there something I'm missing that makes extendable interfaces problematic? Again, in my eyes, the problem is with `interface class`, which allows outside libraries to construct it.

---

I have a proposal that may kill two birds with one stone:
|                    | Inside Library               | Outside Library |
|--------------------|------------------------------|-----------------|
| abstract interface | implement, extend            | implement       |
| interface          | implement, extend, construct | implement       |



The idea is that an interface's constructors would be explictly/implicitly private. It also hinges on the premise that public constructors on interfaces are bad. My proposal removes `class`, which would be too confusing if it could not be constructed, despite being non-abstract. It's also shorter (but not as short as OP's proposal so maybe not quite the full bird). The other bird killed is that the outside library can no longer construct an interface.

**Comment by lrhn** [comment]
I'd be fine with allowing `interface` not followed by `class` to be a shorthand for `abstract interface class`.
(Heck, I've suggested it before myself.)

I don't see any advantage in adding more restrictions than what `abstract interface class` has today.
It prevents you from doing things that you can just choose not to do anyway, and it will only matter inside the same library whether you do it or not. 

Having interfaces with no implementation is not a goal. It's not "using interfaces as intended", because there isn't a single intent for what you can use interfaces for. There is no problem with having implementation, if its actually used. 
The analyzer should be able to tell you if an abstract interface class has implementation, and it's never extended inside the same library. If it doesn't today, it's at least a possibility.

**Comment by mmcdon20** [comment]
> I'd be fine with allowing `interface` not followed by `class` to be a shorthand for `abstract interface class`. (Heck, I've suggested it before myself.)

Well it would at least solve one of my complaints with the existing implementation (too wordy and hard to remember), but I would prefer if it went further than that. 
 
> I don't see any advantage in adding more restrictions than what `abstract interface class` has today. It prevents you from doing things that you can just choose not to do anyway, 

Why have any compile errors? The programmer could simply choose not to do any of the things the compiler would have prevented them from doing. :)

> and it will only matter inside the same library whether you do it or not.

Yes but that is a common scenario, I don't have the data to prove it, but I suspect in a many cases, maybe even most cases, an `abstract interface class` is defined in the same library as classes that implement it.

Take the flutter repo for example, I found 7 different libraries where `abstract interface class` was used and 0 where `interface class` is used. In 5/7 libraries the interfaces were implemented in the same library.

|  interfaces | pure interface? | implemented in same library? | notes |
| --- | --- | --- | --- |
| [CliEnum](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter_tools/lib/src/base/utils.dart#L44) | yes* | no | includes static members |
| [DecorationImagePainter](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter/lib/src/painting/decoration_image.dart#L281) | yes | yes | n/a |
| [ChipAttributes](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter/lib/src/material/chip.dart#L65), [DeletableChipAttributes](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter/lib/src/material/chip.dart#L261), [CheckmarkableChipAttributes](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter/lib/src/material/chip.dart#L348), [SelectableChipAttributes](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter/lib/src/material/chip.dart#L379), [DisabledChipAttributes](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter/lib/src/material/chip.dart#L485), [TappableChipAttributes](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter/lib/src/material/chip.dart#L529)  | yes | yes | n/a |
| [PredictiveBackRoute](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter/lib/src/widgets/routes.dart#L584) | yes | yes | n/a |
| [PlatformLocation](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter_web_plugins/lib/src/navigation_non_web/platform_location.dart#L15) | yes | yes | n/a |
| [HitTestable](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter/lib/src/gestures/hit_test.dart#L18), [HitTestDispatcher](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter/lib/src/gestures/hit_test.dart#L32), [HitTestTarget](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter/lib/src/gestures/hit_test.dart#L38C26-L38C40) | yes | no | n/a |
| [RenderAbstractViewport](https://github.com/flutter/flutter/blob/abbaaea8a12df65ebcb3cb5276e6969cfc3257c3/packages/flutter/lib/src/rendering/viewport.dart#L32C26-L32C48) | yes* | yes | includes static members, also extends `RenderObject` |

> Having interfaces with no implementation is not a goal. It's not "using interfaces as intended", because there isn't a single intent for what you can use interfaces for. 

Surely individual programmers have an intent for what they wish to use their interfaces for, and I suspect a pure interface is the most common intent.

> There is no problem with having implementation, if our actually used. The analyzer should be able to tell you if an abstract interface class has implementation, and it's never extended inside the same library. If it doesn't today, it's at least a possibility.

The analyzer does not do that. Nor are there any lint rules about using an interface in an impure way.

**Comment by gryznar** [comment]
+1. Creating pure interface is cumbersome and very verbose in Dart nowadays. This principle concept should be simplified

**Comment by munificent** [comment]
It would only save you two keywords. That's not nothing, but there is a cost in that if a user sees:

```dart
abstract interface class Foo {}

interface Bar {}
```

They then have to wonder, is there something meaningfully different in how `Foo` is declared versus how `Bar` is declared? The answer would be "no", but they wouldn't necessarily *know* that, so it's a potential source of confusion. Given that the upside is fairly limited, we figured it was best to just have one syntax.

**Comment by gryznar** [comment]
@munificent 
While one syntax avoids confusion, I believe Dart's approach to interfaces is overly verbose. This verbosity is evident in whole language, especially in argument and pure interface declaration, to name a few.
Most developers aren't concerned with the semantic difference between `interface Bar` and `abstract interface class Bar`. Instead, they're confused by the unnecessary complexity of a simple task. Languages like Java, Kotlin and many others offer a clear, concise way to declare pure interface via one keyword which, makes whole process straightforward.
Simplifying this syntax would significantly reduce verbosity and make the language more familiar and accessible to developers coming from other languages. I strongly support this change to improve developer experience and reduce unnecessary boilerplate.

**Comment by mmcdon20** [comment]
@munificent you could add an analyzer warning on

```dart
abstract interface class Foo {}
```

that will recommend simplifying to this form instead

```dart
interface Foo {}
```

**Comment by munificent** [comment]
We could, yes.

The other concern we talked about was the fact that if we allow `interface Foo {}`, it will still let you define concrete methods in it. That's probably unnecessarily confusing because users coming from other languages will expect `interface` to mean "no methods at all".

But there's no reason to have that restriction in Dart. So then we'd have a syntax that is familiar, but sends the wrong signal.

**Comment by mmcdon20** [comment]
@munificent Thats why I also opened [#59466](https://github.com/dart-lang/sdk/issues/59466) and [#55641](https://github.com/dart-lang/sdk/issues/55641) to try and address that.

**Comment by munificent** [comment]
> @munificent Thats why I also opened https://github.com/dart-lang/sdk/issues/59466 and https://github.com/dart-lang/sdk/issues/55641 to try and address that.

The second issue about warning on dead code makes a lot of sense to me. I don't see a motivation for the first one. Why is it bad for an interface class to have concrete code in it?

**Comment by gryznar** [comment]
> Why is it bad for an interface class to have concrete code in it?

Like a default trait implementation in Rust? Also I see this valuable. But still such interface should  be abstract and stateless

**Comment by mmcdon20** [comment]
> > [@munificent](https://github.com/munificent) Thats why I also opened [dart-lang/sdk#59466](https://github.com/dart-lang/sdk/issues/59466) and [dart-lang/sdk#55641](https://github.com/dart-lang/sdk/issues/55641) to try and address that.
> 
> The second issue about warning on dead code makes a lot of sense to me. I don't see a motivation for the first one. Why is it bad for an interface class to have concrete code in it?

The second issue should just be part of regular dead code detection.

The first issue is intended to be a lint rule that users can opt into, not something that is active by default. The reason you would choose to turn the lint rule on, is you want to treat interfaces as an implements only construct. If you intend to use an interface strictly with implements, then there is no point in including concrete code.

**Comment by rubenferreira97** [comment]
> The other concern we talked about was the fact that if we allow `interface Foo {}`, it will still let you define concrete methods in it. That's probably unnecessarily confusing because users coming from other languages will expect `interface` to mean "no methods at all".

[Java](https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html) and [C#](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods) support default methods in interfaces. Isn’t this the same concept you are describing? Java is explicit, it requires `default void myMethod()`, while C# is implicit like Dart if this issue is approved.

**Comment by mmcdon20** [comment]
@rubenferreira97 dart interfaces don't have an equivalent to default methods. You can provide an implementation for a method on an interface in dart but when you implement the interface all of the implementation gets ignored.

```dart
abstract interface class I {
  void a();
  void b() {
    print('b');
  }
}

class C implements I { // error missing concrete implementation for b
  void a() {
    print('a');
  }
}
```

You can use the implementation if you extend the interface which is counter intuitive and can only be done within the same dart library.

```dart
abstract interface class I {
  void a();
  void b() {
    print('b');
  }
}

class C extends I { // works
  void a() {
    print('a');
  }
}
```

---

If you want something like default methods then mixin is a better option than interface in dart.

```dart
mixin M {
  void a();
  void b() {
    print('b');
  }
}

class C with M {
  void a() {
    print('a');
  }
}
```

**Comment by gryznar** [comment]
> @rubenferreira97 dart interfaces don't have an equivalent to default methods. You can provide an implementation for a method on an interface in dart but when you implement the interface all of the implementation gets ignored.

So, what's the point of allowing this at all if it works only for extending?
